### DISCUSSION TITLE
Feature Request: Reinstate Tk in SP-5.40.0

### DISCUSSION THREAD

**Discussion by sisyphus** [StrawberryPerl/Perl-Dist-Strawberry#181] [General]
I've seen quite a few requests for assistance to get Tk installed on SP-5.38 over the last 9 months or so (since SP-5.38.0 was released), and I think there's still more than enough interest in that module to warrant its inclusion in SP-5.40.0.

It's only one file (pTk/mTk/xlib/X11/X.h) that's preventing the build from succeeding - and it's the same issue (and same solution) with both the Tk-804.036 source (from CPAN), and the source on Tk's github repo.

In the hopes of getting this issue fixed (in the CPAN distro and/or the github repo) before the release of perl-5.40.0, I created https://github.com/eserte/perl-tk/issues/104.
But it's looking like that was just a total waste of time.

At least the patch is there if SP developers can see fit to use it.
Or, there's a patched version of X.h downloadable at https://www.perlmonks.org/?node_id=587490;viewmode=public

Cheers,
Rob

**Comment by shawnlaffan** [comment]
(Originally posted to the Tk queue under https://github.com/eserte/perl-tk/issues/104 ).

I've created a patched distribution from the [latest Tk commit](https://github.com/eserte/perl-tk/commit/f0bb386bf2ce9eb9bccb1da8d03321d63865e3d9). For some reason git did not recognise the format so I manually applied them.

I was also unable to update the version to append _001 as it needs to be in several places. If someone can provide a patch for that then that would be helpful.

https://github.com/StrawberryPerl/Perl-Dist-Strawberry/releases/tag/patched_cpan_modules

Whether we include it in the full SP release needs discussion. The compressed blib dir is another 2MB so would increase the downloads commensurately. This is not really very much in the scheme of things these days but it is still non-zero, and I would suspect the majority of users do no need Tk, even though there are many who do.

**Comment by sisyphus** [comment]
Even just having that patched source publicly available is a big improvement.
Having to explain the process of "download-unpack-patch-build-install" to those unfamiliar with that routine is a pain - for both the explainer and the explainee.
But, for now (at least), it's just a matter of telling them to:
```
cpanm https://github.com/StrawberryPerl/Perl-Dist-Strawberry/releases/download/patched_cpan_modules/Tk-804.036_patched.tar.gz
```
As regards the appending of "_001", occurrences of "804.036" are to be found only in the following files:
```
Changes
META.json
META.yml
Tk.pm
Event\Event.pm
```
So, it doesn't look to be a big job.
I'll send a patch later today (to this thread) - once I've written and tested it.

**Comment by sisyphus** [comment]
Not sure if the attached diff will apply cleanly, but it's only a few simple changes that can readily be performed by hand if need be.
Oddly, after unpacking https://github.com/StrawberryPerl/Perl-Dist-Strawberry/releases/download/patched_cpan_modules/Tk-804.036_patched.tar.gz , it seems that Meta.json and Meta.yml have unix line endings, while the other files have windows endings.

Anyway, with the changes specified in the attached diff, Tk-804.036_001 built, tested, and installed just fine on SP-5.39.10.
After installation, I get:
```
> perl -MTk -le "print $Tk::VERSION;"
804.036001
```
I was actually expecting to see "804.036_001" but, hopefully, it's just that my expectations are outdated ;-)

[diff.txt](https://github.com/StrawberryPerl/Perl-Dist-Strawberry/files/15208029/diff.txt)

**Comment by shawnlaffan** [comment]
Thanks Rob.  The patch again did not apply cleanly using git.  It was easy enough to do manually, as you noted, so I did that.  

The updated file is at https://github.com/StrawberryPerl/Perl-Dist-Strawberry/releases/download/patched_cpan_modules/Tk-804.036_001.tar.gz

I deleted the previous ```tar.gz``` file to reduce confusion (I hope).

**Comment by shawnlaffan** [comment]
A patched version of Tk can be installed using the instructions at https://github.com/StrawberryPerl/Perl-Dist-Strawberry/issues/87#issuecomment-2292839449
